File: apps/frontend/src/features/source-control/schemas.ts
import { z } from "zod";

export const SourceControlMergeStateSchema = z.enum([
  "merged",
  "ready",
  "diverged",
  "behind",
  "unknown",
]);

export const SourceControlChangedFileSchema = z.object({
  path: z.string().min(1),
  code: z.string().min(1),
  staged: z.boolean().default(false),
  unstaged: z.boolean().default(false),
});

export const SourceControlCommitPreviewSchema = z.object({
  hash: z.string().nullable().optional(),
  subject: z.string().min(1),
  body: z.string().nullable().optional(),
  authorName: z.string().nullable().optional(),
  authoredAt: z.string().datetime({ offset: true }).nullable().optional(),
});

export const SourceControlCommitActivityEntrySchema = z.object({
  hash: z.string().min(1),
  subject: z.string().min(1),
  authoredAt: z.string().datetime({ offset: true }),
  url: z.string().url().nullable().optional(),
});

export const SourceControlBranchPreviewSchema = z.object({
  name: z.string().min(1),
  isActive: z.boolean().default(false),
  ahead: z.number().int().min(0).default(0),
  behind: z.number().int().min(0).default(0),
  mergedIntoBase: z.boolean().nullable().optional(),
  mergeState: SourceControlMergeStateSchema,
});

export const SourceControlMergePreviewEntrySchema = z.object({
  branch: z.string().min(1),
  mergeState: SourceControlMergeStateSchema,
  ahead: z.number().int().min(0).default(0),
  behind: z.number().int().min(0).default(0),
});

export const SourceControlWorktreeEntrySchema = z.object({
  path: z.string().min(1),
  branch: z.string().nullable().optional(),
  isCurrent: z.boolean().default(false),
});

export const SourceControlBotSyncEntrySchema = z.object({
  botName: z.string().min(1),
  botStatus: z.enum(["idle", "active"]),
  runtime: z.string().min(1),
  matchedBranch: z.string().nullable().optional(),
  inSync: z.boolean().default(false),
  branchCandidates: z.array(z.string()).default([]),
  source: z.enum(["agent", "snapshot"]).default("agent"),
  snapshotName: z.string().nullable().optional(),
  sessionCount: z.number().int().min(0).default(0),
});

export const SourceControlPullRequestSchema = z.object({
  number: z.number().int().positive(),
  title: z.string().min(1),
  state: z.enum(["open", "merged", "closed"]).default("open"),
  headBranch: z.string().min(1),
  baseBranch: z.string().min(1),
  url: z.string().url().nullable().optional(),
  author: z.string().nullable().optional(),
  isDraft: z.boolean().default(false),
});

export const SourceControlFailedCheckSchema = z.object({
  name: z.string().min(1),
  workflowName: z.string().nullable().optional(),
  conclusion: z.string().min(1).default("unknown"),
  detailsUrl: z.string().url().nullable().optional(),
});

export const SourceControlReviewFeedbackEntrySchema = z.object({
  source: z.enum(["issue_comment", "review", "review_comment"]).default("review"),
  content: z.string().min(1),
  state: z.string().nullable().optional(),
  author: z.string().nullable().optional(),
  filePath: z.string().nullable().optional(),
  submittedAt: z.string().datetime({ offset: true }).nullable().optional(),
  url: z.string().url().nullable().optional(),
});

export const SourceControlPullRequestDiagnosticsSchema = z.object({
  pullRequest: SourceControlPullRequestSchema,
  mergeable: z.string().nullable().optional(),
  mergeStateStatus: z.string().nullable().optional(),
  hasMergeConflicts: z.boolean().default(false),
  failedChecks: z.array(SourceControlFailedCheckSchema).default([]),
  feedback: z.array(SourceControlReviewFeedbackEntrySchema).default([]),
});

export const SourceControlReviewContentSchema = z.object({
  kind: z.enum(["review", "comment", "decision"]),
  content: z.string().min(1),
  state: z.string().nullable().optional(),
  author: z.string().nullable().optional(),
  submittedAt: z.string().datetime({ offset: true }).nullable().optional(),
  url: z.string().url().nullable().optional(),
});

export const SourceControlPreviewResponseSchema = z.object({
  repositoryRoot: z.string().min(1),
  projectPath: z.string().nullable().optional(),
  activeBranch: z.string().min(1),
  baseBranch: z.string().min(1),
  dirty: z.boolean(),
  refreshedAt: z.string().datetime({ offset: true }),
  changedFiles: z.array(SourceControlChangedFileSchema).default([]),
  commitPreview: SourceControlCommitPreviewSchema,
  branches: z.array(SourceControlBranchPreviewSchema).default([]),
  mergePreview: z.array(SourceControlMergePreviewEntrySchema).default([]),
  worktrees: z.array(SourceControlWorktreeEntrySchema).default([]),
  gxBots: z.array(SourceControlBotSyncEntrySchema).default([]),
  pullRequests: z.array(SourceControlPullRequestSchema).default([]),
  conflictedPullRequests: z.array(SourceControlPullRequestDiagnosticsSchema).default([]),
  botFeedbackPullRequests: z.array(SourceControlPullRequestDiagnosticsSchema).default([]),
  quickActions: z.array(z.string()).default([]),
});

export const SourceControlCommitActivityResponseSchema = z.object({
  repositoryRoot: z.string().min(1),
  projectPath: z.string().nullable().optional(),
  commits: z.array(SourceControlCommitActivityEntrySchema).default([]),
});

export const SourceControlBranchDetailsResponseSchema = z.object({
  repositoryRoot: z.string().min(1),
  projectPath: z.string().nullable().optional(),
  branch: z.string().min(1),
  baseBranch: z.string().min(1),
  mergeState: SourceControlMergeStateSchema,
  ahead: z.number().int().min(0).default(0),
  behind: z.number().int().min(0).default(0),
  changedFiles: z.array(SourceControlChangedFileSchema).default([]),
  linkedBots: z.array(z.string()).default([]),
  pullRequest: SourceControlPullRequestSchema.nullable().optional(),
  reviewContent: SourceControlReviewContentSchema.nullable().optional(),
});

export const SourceControlCreatePullRequestResponseSchema = z.object({
  status: z.literal("created"),
  branch: z.string().min(1),
  baseBranch: z.string().min(1),
  pullRequest: SourceControlPullRequestSchema.nullable().optional(),
  message: z.string().min(1),
});

export const SourceControlMergePullRequestResponseSchema = z.object({
  status: z.literal("merged"),
  branch: z.string().min(1),
  pullRequestNumber: z.number().int().positive().nullable().optional(),
  message: z.string().min(1),
});

export const SourceControlDeleteBranchResponseSchema = z.object({
  status: z.literal("deleted"),
  branch: z.string().min(1),
  message: z.string().min(1),
});

export type SourceControlPreviewResponse = z.infer<typeof SourceControlPreviewResponseSchema>;
export type SourceControlCommitActivityResponse = z.infer<typeof SourceControlCommitActivityResponseSchema>;
export type SourceControlCommitActivityEntry = z.infer<typeof SourceControlCommitActivityEntrySchema>;
export type SourceControlMergeState = z.infer<typeof SourceControlMergeStateSchema>;
export type SourceControlBranchDetailsResponse = z.infer<typeof SourceControlBranchDetailsResponseSchema>;

File: app/modules/source_control/schemas.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import Field

from app.modules.shared.schemas import DashboardModel

MergeState = Literal["merged", "ready", "diverged", "behind", "unknown"]
BotStatus = Literal["idle", "active"]
PullRequestState = Literal["open", "merged", "closed"]
ReviewContentKind = Literal["review", "comment", "decision"]
CheckConclusion = Literal[
    "failure",
    "timed_out",
    "cancelled",
    "action_required",
    "startup_failure",
    "stale",
    "neutral",
    "success",
    "skipped",
    "pending",
    "unknown",
]


class SourceControlChangedFile(DashboardModel):
    path: str
    code: str
    staged: bool = False
    unstaged: bool = False


class SourceControlCommitPreview(DashboardModel):
    hash: str | None = None
    subject: str
    body: str | None = None
    author_name: str | None = None
    authored_at: datetime | None = None


class SourceControlCommitActivityEntry(DashboardModel):
    hash: str
    subject: str
    authored_at: datetime
    url: str | None = None


class SourceControlBranchPreview(DashboardModel):
    name: str
    is_active: bool = False
    ahead: int = 0
    behind: int = 0
    merged_into_base: bool | None = None
    merge_state: MergeState = "unknown"


class SourceControlMergePreviewEntry(DashboardModel):
    branch: str
    merge_state: MergeState
    ahead: int = 0
    behind: int = 0


class SourceControlWorktreeEntry(DashboardModel):
    path: str
    branch: str | None = None
    is_current: bool = False


class SourceControlBotSyncEntry(DashboardModel):
    bot_name: str
    bot_status: BotStatus
    runtime: str
    matched_branch: str | None = None
    in_sync: bool = False
    branch_candidates: list[str] = Field(default_factory=list)
    source: Literal["agent", "snapshot"] = "agent"
    snapshot_name: str | None = None
    session_count: int = 0


class SourceControlReviewContent(DashboardModel):
    kind: ReviewContentKind
    content: str
    state: str | None = None
    author: str | None = None
    submitted_at: datetime | None = None
    url: str | None = None


class SourceControlPullRequestPreview(DashboardModel):
    number: int
    title: str
    state: PullRequestState = "open"
    head_branch: str
    base_branch: str
    url: str | None = None
    author: str | None = None
    is_draft: bool = False


class SourceControlFailedCheck(DashboardModel):
    name: str
    workflow_name: str | None = None
    conclusion: CheckConclusion = "unknown"
    details_url: str | None = None


class SourceControlReviewFeedbackEntry(DashboardModel):
    source: Literal["issue_comment", "review", "review_comment"] = "review"
    content: str
    state: str | None = None
    author: str | None = None
    file_path: str | None = None
    submitted_at: datetime | None = None
    url: str | None = None


class SourceControlPullRequestDiagnostics(DashboardModel):
    pull_request: SourceControlPullRequestPreview
    mergeable: str | None = None
    merge_state_status: str | None = None
    has_merge_conflicts: bool = False
    failed_checks: list[SourceControlFailedCheck] = Field(default_factory=list)
    feedback: list[SourceControlReviewFeedbackEntry] = Field(default_factory=list)


class SourceControlPreviewResponse(DashboardModel):
    repository_root: str
    project_path: str | None = None
    active_branch: str
    base_branch: str
    dirty: bool
    refreshed_at: datetime
    changed_files: list[SourceControlChangedFile] = Field(default_factory=list)
    commit_preview: SourceControlCommitPreview
    branches: list[SourceControlBranchPreview] = Field(default_factory=list)
    merge_preview: list[SourceControlMergePreviewEntry] = Field(default_factory=list)
    worktrees: list[SourceControlWorktreeEntry] = Field(default_factory=list)
    gx_bots: list[SourceControlBotSyncEntry] = Field(default_factory=list)
    pull_requests: list[SourceControlPullRequestPreview] = Field(default_factory=list)
    conflicted_pull_requests: list[SourceControlPullRequestDiagnostics] = Field(default_factory=list)
    bot_feedback_pull_requests: list[SourceControlPullRequestDiagnostics] = Field(default_factory=list)
    quick_actions: list[str] = Field(default_factory=list)


class SourceControlCommitActivityResponse(DashboardModel):
    repository_root: str
    project_path: str | None = None
    commits: list[SourceControlCommitActivityEntry] = Field(default_factory=list)


class SourceControlBranchDetailsResponse(DashboardModel):
    repository_root: str
    project_path: str | None = None
    branch: str
    base_branch: str
    merge_state: MergeState
    ahead: int = 0
    behind: int = 0
    changed_files: list[SourceControlChangedFile] = Field(default_factory=list)
    linked_bots: list[str] = Field(default_factory=list)
    pull_request: SourceControlPullRequestPreview | None = None
    review_content: SourceControlReviewContent | None = None


class SourceControlCreatePullRequestRequest(DashboardModel):
    project_id: str | None = None
    branch: str
    base_branch: str | None = None
    title: str | None = None
    body: str | None = None
    draft: bool = False


class SourceControlCreatePullRequestResponse(DashboardModel):
    status: Literal["created"]
    branch: str
    base_branch: str
    pull_request: SourceControlPullRequestPreview | None = None
    message: str


class SourceControlMergePullRequestRequest(DashboardModel):
    project_id: str | None = None
    branch: str
    pull_request_number: int | None = None
    base_branch: str | None = None
    delete_branch: bool = True
    squash: bool = False


class SourceControlMergePullRequestResponse(DashboardModel):
    status: Literal["merged"]
    branch: str
    pull_request_number: int | None = None
    message: str


class SourceControlDeleteBranchRequest(DashboardModel):
    project_id: str | None = None
    branch: str


class SourceControlDeleteBranchResponse(DashboardModel):
    status: Literal["deleted"]
    branch: str
    message: str
